Add documentation for Performance Counters (#427)
ResearchComputing/Documentation: docs/programming/profiling-nvidia-gpu-performance.md

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|  No running processes found                                                             |
+-----------------------------------------------------------------------------------------+

```
#### What’s Different with MIG?

- **MIG Mode:** You'll see ```MIG M.: Enabled``` in the main GPU listing.

- **GPU Utilization:** The parent GPU will often show ```N/A``` for ```GPU-Util```; usage is tracked per MIG instance instead.

- **MIG Devices Section:** This shows each available MIG instance, how much memory and compute it has, and its usage stats.

- **Process Table:** When jobs are running, this table will show which process is attached to which MIG slice.

> ### Monitoring GPU Usage While a Job is Running

Reviewer: ```suggestion
### Manually Monitoring GPU Usage While a Job is Running
```

To check performance while your job is active:

1. Submit or start your GPU job

>  - You can run your GPU job either interactively (e.g. using ```sinteractive```) or via a batch script submitted with ```sbatch```

Reviewer: ```suggestion
     - You can run your GPU job either interactively (e.g. using `sinteractive`) or via a batch script submitted with `sbatch`
```

2. Identify the node where your job is running

>  - When the job is running, use the squeue command to find which compute node it's on:

Reviewer: ```suggestion
     - When the job is running, use the squeue command to find which compute node it's on:
```

      ``` 
      $ squeue -u $USER 
      ```

>  - Look for the NODELIST(REASON) column, this shows the name of the compute node assigned to your job (e.g. ```c3gpu-c2-u13```).

Reviewer: ```suggestion
     - Look for the NODELIST(REASON) column, this shows the name of the compute node assigned to your job (e.g. `c3gpu-c2-u13`).
```

3. ```ssh``` into the compute node

>  - Once you have the node name, ```ssh``` into it directly:

Reviewer: ```suggestion
     - Once you have the node name, `ssh` into it directly:
```

      ```
      $ ssh <node-name>
      # Example:
      $ ssh c3gpu-c2-u13
      ```

      ```{note} 
      You can only ```ssh``` into a compute node if you have an active job running on it. 
      ```

4. Run ```nvidia-smi``` 

> - Once logged into the node, use ```nvidia-smi``` command to monitor GPU usage.

Reviewer: ````suggestion
    - Once logged into the node, you can use the `nvidia-smi` command to monitor GPU usage. 
    - You can alternatively use:
    ```
    $ nvidia-smi -l 2
    ```
   where `-l 2` is a flag that will refresh the output every 2 seconds. You can change the interval by adjusting the number (`-l 5` for every 5 seconds). Press `Ctrl+C` to stop the loop.
````

Reviewer: Please check the formatting here. I did this within GitHub.

> ### Real-Time Monitoring of GPU Usage
> 
> To actively monitor GPU performance during a running job, you can use either the ```watch``` command or the built-in looping functionality of ```nvidia-smi```.
> 
> **Option 1:** Using the ```watch``` repeatedly executes ```nvidia-smi``` at regular intervals and highlights changes in output, making it easier to spot fluctuations in GPU load or memory usage.
> 
> ```
> $ watch -n 2 nvidia-smi
> ```
> Here, ```-n 2``` sets the refresh interval to every 2 seconds (you can increase or decrease this number as needed).
> 
> ```{note}
> Each refresh with ```watch``` creates a new process, which can add some system overhead over time.
> ```
> 
> **Option 2**: A more efficient alternative is to use the ```-l``` (loop) option built into ```nvidia-smi```. This avoids spawning new processes by refreshing internally. 
> ```
> $ nvidia-smi -l 2
> ```
> The ```-l 2``` flag will refresh the output every 2 seconds. You can change the interval by adjusting the number (```-l 5``` for every 5 seconds). Press ```Ctrl+C``` to stop the loop.

Reviewer: ````suggestion
### Monitoring GPU Usage within a Job Script

To actively monitor GPU performance within a job script, you can run `nvidia-smi` as a background process and redirect its output. Below we provide an example script that redirects the `nvidia-smi` output to the output file `gpu_usage-jobid-${SLURM_JOB_ID}.log` every 60 seconds, where `${SLURM_JOB_ID}` is an environment variable specifying the Job ID. 

```
#!/bin/bash
#SBATCH --nodes=1
#SBATCH --ntasks=10
#SBATCH --gres=gpu
#SBATCH --partition=atesting_a100 
#SBATCH --qos=testing
#SBATCH --time=00:10:00

# Start GPU monitoring in the background
nvidia-smi -l 60 >> gpu_usage-jobid-${SLURM_JOB_ID}.log &
MONITOR_PID=$!

# Replace the sleep command with your own command e.g. "python my_script.py"
# Here we just sleep for 100 seconds
sleep 100

# Stop monitoring after job finishes
kill $MONITOR_PID
```
````

I changed it to the above to cut down on repetition and highlight how to do this in a job script, which may not be obvious to some people.

## Nsight Compute (ncu)

NVIDIA Nsight Compute is a command-line CUDA kernel profiler that provides detailed performance metrics for GPU kernels, helping users identify, and resolve bottlenecks in CUDA applications. It is particularly well-suited for low-level kernel analysis and optimization.

### Why These Metrics Matter?

Modern GPUs, such as the NVIDIA A100, have hundreds of compute units (Streaming Multiprocessors or SMs). To fully exploit this parallel architecture, your kernel must be configured to launch enough threads and blocks to keep these units busy.

Key Features:

- Collects performance data on SM utilization, memory throughput, warp execution efficiency, and more.

- Offers optimization guidance via diagnostic messages.

- Enables deep inspection of stalls, occupancy, and instruction efficiency.

```{caution}
Collecting performance data using ```ncu``` can incur significant runtime overhead. For production runs, disable profiling. See [Nsight Compute Overhead](https://docs.nvidia.com/nsight-compute/ProfilingGuide/index.html#overhead) for more details.
```

### Getting Started

To use ```ncu```, first load the appropriate CUDA module:

```
$ module load cuda
```

> Compile the CUDA application using: 

Reviewer: Please point people to the section where you specified `vectorAdd.cu`

```
$ nvcc -o vectorAdd vectorAdd.cu
```

Next to invoke ```ncu```, prefix it to your compiled CUDA application:

```
$ ncu --set full --target-processes all ./vectorAdd
```
- ```--set full```: Collects a comprehensive set of performance metrics.

- ```--target-processes all```: Profiles all child processes (useful for multi-threaded applications).

```{note}
```ncu``` is not compatible with MIG-enabled GPUs. Ensure you run ```ncu``` only on A100 nodes without MIG.
```

> ::::{dropdown} Click here to view full ```ncu``` report

Reviewer: ```suggestion
::::{dropdown} Click here to view the full `ncu` report
```

```{tip}

| Symptom               | Likely Cause                          | Recommended Fix                           | 
| :----------------- | :------------------------------------------- |:-------------------------------------------------- |
| Grid size is smaller than #SMs	| Not enough blocks to occupy all compute units	| Increase the number of blocks by enlarging the problem size |
| SM throughput is very low (< 10%)	| GPU is mostly idle	| Optimize workload or parallelism granularity |
| Waves Per SM is 0	| Too little parallel work	| Increase N or restructure the launch configuration |
```

In the original kernel, the launch configuration was:
```
#define N 1024

dim3 block(256);
dim3 grid((N + block.x - 1) / block.x);  // → grid.x = 4 blocks
```
This launches just 4 blocks, causing underutilization.
To improve utilization, update the launch configuration to the following:
```
#define N (1 << 20)  // 1,048,576 elements

dim3 block(256);
dim3 grid((N + block.x - 1) / block.x);  // → grid.x = 4096 blocks
```
This updated configuration allows the GPU to schedule multiple waves of work across all SMs, leading to much better throughput and performance.

> ::::{dropdown} Click here to view ```ncu``` report for updated configuration

Reviewer: ```suggestion
::::{dropdown} Click here to view the `ncu` report for the updated configuration
```

          or workload imbalances during the kernel execution. Load imbalances can occur between warps within a block
          as well as across blocks of the same kernel. See the CUDA Best Practices Guide
          (https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/index.html#occupancy) for more details on
          optimizing occupancy.

    Section: Source Counters
    ------------------------- ----------- ------------
    Metric Name               Metric Unit Metric Value
    ------------------------- ----------- ------------
    Branch Instructions Ratio           %         0.12
    Branch Instructions              inst       65,536
    Branch Efficiency                   %            0
    Avg. Divergent Branches                          0
    ------------------------- ----------- ------------

::::

## Nsight Systems (nsys)

Nsight Systems  is a system-wide profiler that traces the interactions between CPU and GPU, memory transfers, and OS-level activity. It’s useful for understanding:

- Launch latency of GPU kernels.

- Investigate host-device memory transfers.

- Identify CPU-side inefficiencies (e.g., blocking, thread stalls)

### Getting Started

> To use ```nsys```, load the appropriate CUDA module, compile and invoke ```nsys``` by prefixing it to your CUDA application:

Reviewer: Please tell them where they can find `vectorAdd.cu`

```
$ module load cuda
$ nvcc -o vectorAdd vectorAdd.cu
$ nsys profile --trace=cuda,osrt ./vectorAdd
```
- ```--trace=cuda```: Capture CUDA kernel launches and memory transfers.

- ```--trace=osrt```: Trace OS runtime activity (threads, processes, etc.).

This generates a report file in binary format:
```
./report.nsys-rep
```
To view a quick CLI summary of the captured data:
```
$ nsys stats report.nsys-rep
```

> ::::{dropdown} Click here to view full output 

Reviewer: ```suggestion
::::{dropdown} Click here to view the full output 
```

    cudaMemcpy(C, d_C, size, cudaMemcpyDeviceToHost);

    delete[] A; delete[] B; delete[] C;
    cudaFree(d_A); cudaFree(d_B); cudaFree(d_C);
    return 0;
}
```

## nvidia-smi

`nvidia-smi` (System Management Interface) is a command-line utility that provides real-time information about GPU utilization, memory usage, temperature, and running processes.

### Getting Started

```
$ nvidia-smi
```
Example output of `nvidia-smi` on the `aa100` partition
```
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 570.124.06             Driver Version: 570.124.06     CUDA Version: 12.8     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA A100-PCIE-40GB          Off |   00000000:21:00.0 Off |                    0 |

> | N/A   33C    P0             40W /  250W |       1MiB /  40960MiB |      2%      Default |

Reviewer: These values do not make sense to me (primarily the memory one). I think you only added the process to the table, but did not update these values. Please update these values so they make sense. Additionally, update the table in the `Sensor and Resource Usage Metrics` subsection.

Here’s a simple CUDA program `vectorAdd.cu` that adds two vectors of floats.

```
#include <iostream>
#include <cuda_runtime.h>

#define N 1024  // Size of the vectors

__global__ void vectorAdd(float *A, float *B, float *C) {
    int idx = threadIdx.x + blockIdx.x * blockDim.x;
    if (idx < N) {
        C[idx] = A[idx] + B[idx];
    }
}

int main() {
    float *A = new float[N], *B = new float[N], *C = new float[N];
    for (int i = 0; i < N; i++) {
        A[i] = static_cast<float>(rand()) / RAND_MAX;
        B[i] = static_cast<float>(rand()) / RAND_MAX;
    }

    float *d_A, *d_B, *d_C;
    size_t size = N * sizeof(float);
    cudaMalloc(&d_A, size); cudaMalloc(&d_B, size); cudaMalloc(&d_C, size);
    cudaMemcpy(d_A, A, size, cudaMemcpyHostToDevice);
    cudaMemcpy(d_B, B, size, cudaMemcpyHostToDevice);

    dim3 block(256);
    dim3 grid((N + block.x - 1) / block.x);
    vectorAdd<<<grid, block>>>(d_A, d_B, d_C);
    cudaMemcpy(C, d_C, size, cudaMemcpyDeviceToHost);

    delete[] A; delete[] B; delete[] C;
    cudaFree(d_A); cudaFree(d_B); cudaFree(d_C);
    return 0;
}
```

> ## nvidia-smi

Reviewer: ```suggestion
## `nvidia-smi`
```

On another heading you use the back ticks to indicate a command. For consistency sake, I would either do that here or remove the back ticks from the other heading. I am fine with either.

Author: Thank you for your suggestion! I’ve removed the backticks from the other heading to maintain consistency, as Nsight Compute and Nsight Systems don’t use backticks for titles.

`nvidia-smi` (System Management Interface) is a command-line utility that provides real-time information about GPU utilization, memory usage, temperature, and running processes.

### Getting Started

```
$ nvidia-smi
```

> Example output of `nvidia-smi` on the `aa100` patition

Reviewer: ```suggestion
Example output of `nvidia-smi` on the `aa100` partition
```

- **GPU Utilization:** The parent GPU will often show `N/A` for `GPU-Util`; usage is tracked per MIG instance instead.

- **MIG Devices Section:** This shows each available MIG instance, how much memory and compute it has, and its usage stats.

- **Process Table:** When jobs are running, this table will show which process is attached to which MIG slice.

```{note}
When using MIG-enabled GPUs, refer to the MIG devices section for information about individual MIG slices. Do not rely on the main GPU table for MIG-specific details.
```

### Manually Monitoring GPU Usage While a Job is Running

To check performance while your job is active:

1. Submit or start your GPU job

     - You can run your GPU job either interactively (e.g. using `sinteractive`) or via a batch script submitted with `sbatch`

2. Identify the node where your job is running

     - When the job is running, use the squeue command to find which compute node it's on:

      ``` 
      $ squeue -u $USER 
      ```

     - Look for the NODELIST(REASON) column, this shows the name of the compute node assigned to your job (e.g. `c3gpu-c2-u13`).

3. ```ssh``` into the compute node

>      - Once you have the node name, `ssh` into it directly:
> 
>       ```
>       $ ssh <node-name>
>       # Example:
>       $ ssh c3gpu-c2-u13
>       ```

Reviewer: ````suggestion
     - Once you have the node name, `ssh` into it directly. For example, if your GPU job was running on `c3gpu-c2-u13`, you would do the following:

      ```
      $ ssh c3gpu-c2-u13
      ```
````

| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0    0    0           110318      C    ./vectorAdd                            34MiB  |
+-----------------------------------------------------------------------------------------+

```
| Column               | Value                                         | 
| :----------------- | :-------------------------------------------------- | 
| GPU | In this case, the GPU being used is GPU 0. | 
| GI/CI	| 0 / 0 — Since MIG (Multi-Instance GPU) is not in use here, both are listed as 0. | 
| PID	| 110318 — This is the unique Process ID of the application using the GPU. | 
| Type	| C — This indicates that the process is a Compute process, meaning the GPU is being used for calculations or data processing.| 
| Process Name	| ./vectorAdd — This is the name of the executable using the GPU. | 

> | GPU Memory Usage	| 34MiB — This shows that the process is using 34 MiB of memory.| 

Reviewer: Thank you very much for adding an example where there is a process running. This does feel out of place though for me. I think there are two options you can go with here:
1. Keep the example in it's entirety, but make it a subsection under `Getting Started`, and before the section `nvidia-smi on MIG-Enabled GPUs`
2. Modify your `nvidia-smi` output under the `Getting Started` section so that it has a process in it (you would then need to modify the tables accordingly).

```

## Nsight Compute (ncu)

NVIDIA Nsight Compute is a command-line CUDA kernel profiler that provides detailed performance metrics for GPU kernels, helping users identify, and resolve bottlenecks in CUDA applications. It is particularly well-suited for low-level kernel analysis and optimization.

### Why These Metrics Matter?

Modern GPUs, such as the NVIDIA A100, have hundreds of compute units (Streaming Multiprocessors or SMs). To fully exploit this parallel architecture, your kernel must be configured to launch enough threads and blocks to keep these units busy.

Key Features:

- Collects performance data on SM utilization, memory throughput, warp execution efficiency, and more.

- Offers optimization guidance via diagnostic messages.

- Enables deep inspection of stalls, occupancy, and instruction efficiency.

```{caution}
Collecting performance data using `ncu` can incur significant runtime overhead. For production runs, disable profiling. See [Nsight Compute Overhead](https://docs.nvidia.com/nsight-compute/ProfilingGuide/index.html#overhead) for more details.
```

### Getting Started

To use `ncu`, first load the appropriate CUDA module:

```
$ module load cuda
```

> Compile the CUDA code (`vectorAdd.cu`), provided in [Sample CUDA Code: Vector Addition](../programming/profiling-nvidia-gpu-performance.md#sample-cuda-code-vector-addition) section, using `nvcc` compiler:

Reviewer: ```suggestion
Compile the CUDA code (`vectorAdd.cu`), provided in [Sample CUDA Code: Vector Addition](../programming/profiling-nvidia-gpu-performance.md#sample-cuda-code-vector-addition) section, using the `nvcc` compiler:
```

```
$ nvcc -o vectorAdd vectorAdd.cu
```

> Next to invoke `ncu`, prefix it to your compiled CUDA application:

Reviewer: ```suggestion
Next, to invoke `ncu`, prefix it to your compiled CUDA application:
```

        B[i] = static_cast<float>(rand()) / RAND_MAX;
    }

    float *d_A, *d_B, *d_C;
    cudaMalloc(&d_A, size);
    cudaMalloc(&d_B, size);
    cudaMalloc(&d_C, size);

    cudaMemcpy(d_A, A, size, cudaMemcpyHostToDevice);
    cudaMemcpy(d_B, B, size, cudaMemcpyHostToDevice);

    dim3 block(16, 16);
    dim3 grid((N + block.x - 1) / block.x, (N + block.y - 1) / block.y);

    auto start = std::chrono::high_resolution_clock::now();

    matrixMul<<<grid, block>>>(d_A, d_B, d_C);
    cudaDeviceSynchronize();

    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::minutes>(end - start);

    cudaMemcpy(C, d_C, size, cudaMemcpyDeviceToHost);

    std::cout << "Execution time: " << duration.count() << " minutes" << std::endl;

    delete[] A;
    delete[] B;
    delete[] C;
    cudaFree(d_A);
    cudaFree(d_B);
    cudaFree(d_C);

    return 0;
}

```
````
`````

## nvidia-smi

`nvidia-smi` (System Management Interface) is a command-line utility that provides real-time information about GPU utilization, memory usage, temperature, and running processes. 

### Getting Started

```
$ nvidia-smi
```

> Example output of `nvidia-smi` on the `aa100` partition for [Matrix Multiplication](?tabset-cuda-code=cuda-matrix-multiply#tabset-ref-cuda-code){.external} code 

Reviewer: ```suggestion
Example output of `nvidia-smi` on the `aa100` partition using [matrixMultiply.cu](?tabset-cuda-code=cuda-matrix-multiply#tabset-ref-cuda-code){.external}:
```

|                  |                                  |        ECC|                       |
|==================+==================================+===========+=======================|
|  0    1   0   0  |             107MiB / 20096MiB    | 42      0 |  3   0    2    0    0 |
|                  |                 0MiB / 32767MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|  No running processes found                                                             |
+-----------------------------------------------------------------------------------------+

```

> #### What’s Different with MIG?

Reviewer: ```suggestion
#### What’s different with MIG?
```

### Monitoring GPU Usage within a Job Script

To actively monitor GPU performance within a job script, you can run `nvidia-smi` as a background process and redirect its output. Below we provide an example script that redirects the `nvidia-smi` output to the output file `gpu_usage-jobid-${SLURM_JOB_ID}.log` every 60 seconds, where `${SLURM_JOB_ID}` is an environment variable specifying the Job ID. 

```
#!/bin/bash
#SBATCH --nodes=1
#SBATCH --ntasks=10
#SBATCH --gres=gpu
#SBATCH --partition=atesting_a100 
#SBATCH --qos=testing
#SBATCH --time=00:10:00

# Start GPU monitoring in the background
nvidia-smi -l 60 >> gpu_usage-jobid-${SLURM_JOB_ID}.log &
MONITOR_PID=$!

# Replace the sleep command with your own command e.g. "python my_script.py"
# Here we just sleep for 100 seconds
sleep 100

# Stop monitoring after job finishes
kill $MONITOR_PID
```

## Nsight Compute (ncu)

NVIDIA Nsight Compute is a command-line CUDA kernel profiler that provides detailed performance metrics for GPU kernels, helping users identify, and resolve bottlenecks in CUDA applications. It is particularly well-suited for low-level kernel analysis and optimization.

> ### Why These Metrics Matter?

Reviewer: ```suggestion
### Why do these metrics matter?
```

Modern GPUs, such as the NVIDIA A100, have hundreds of compute units (Streaming Multiprocessors or SMs). To fully exploit this parallel architecture, your kernel must be configured to launch enough threads and blocks to keep these units busy.

Key Features:

- Collects performance data on SM utilization, memory throughput, warp execution efficiency, and more.

- Offers optimization guidance via diagnostic messages.

- Enables deep inspection of stalls, occupancy, and instruction efficiency.

```{caution}
Collecting performance data using `ncu` can incur significant runtime overhead. For production runs, disable profiling. See [Nsight Compute Overhead](https://docs.nvidia.com/nsight-compute/ProfilingGuide/index.html#overhead) for more details.
```

### Getting Started

To use `ncu`, first load the appropriate CUDA module:

```
$ module load cuda
```

> Compile the CUDA code (`vectorAdd.cu`), provided in [Sample CUDA code](?tabset-cuda-code=cuda-vector-add#tabset-ref-cuda-code){.external} section, using the `nvcc` compiler:

Reviewer: ```suggestion
Compile the CUDA code [vectorAdd.cu](?tabset-cuda-code=cuda-vector-add#tabset-ref-cuda-code){.external}, using the `nvcc` compiler:
```

          or workload imbalances during the kernel execution. Load imbalances can occur between warps within a block
          as well as across blocks of the same kernel. See the CUDA Best Practices Guide
          (https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/index.html#occupancy) for more details on
          optimizing occupancy.

    Section: Source Counters
    ------------------------- ----------- ------------
    Metric Name               Metric Unit Metric Value
    ------------------------- ----------- ------------
    Branch Instructions Ratio           %         0.12
    Branch Instructions              inst       65,536
    Branch Efficiency                   %            0
    Avg. Divergent Branches                          0
    ------------------------- ----------- ------------

::::

## Nsight Systems (nsys)

Nsight Systems  is a system-wide profiler that traces the interactions between CPU and GPU, memory transfers, and OS-level activity. It’s useful for understanding:

- Launch latency of GPU kernels.

- Investigate host-device memory transfers.

- Identify CPU-side inefficiencies (e.g., blocking, thread stalls)

### Getting Started

> To use `nsys`, first load the appropriate CUDA module. Then compile the CUDA code (`vectorAdd.cu`), available in the [Sample CUDA code](?tabset-cuda-code=cuda-vector-add#tabset-ref-cuda-code){.external} section, using the `nvcc` compiler. Finally, run `nsys` by prefixing it to your compiled application.

Reviewer: ```suggestion
To use `nsys`, first load the appropriate CUDA module. Then compile the CUDA code [vectorAdd.cu](?tabset-cuda-code=cuda-vector-add#tabset-ref-cuda-code){.external} using the `nvcc` compiler. Finally, run `nsys` by prefixing it to your compiled application.
```

Processing [report.sqlite] with [/curc/sw/cuda/12.1.1/nsight-systems-2023.1.2/host-linux-x64/reports/cuda_gpu_mem_time_sum.py]...

 ** CUDA GPU MemOps Summary (by Time) (cuda_gpu_mem_time_sum):

 Time (%)  Total Time (ns)  Count  Avg (ns)  Med (ns)  Min (ns)  Max (ns)  StdDev (ns)      Operation
 --------  ---------------  -----  --------  --------  --------  --------  -----------  ------------------
     52.2            2,656      2   1,328.0   1,328.0     1,216     1,440        158.4  [CUDA memcpy HtoD]
     47.8            2,432      1   2,432.0   2,432.0     2,432     2,432          0.0  [CUDA memcpy DtoH]

Processing [report.sqlite] with [/curc/sw/cuda/12.1.1/nsight-systems-2023.1.2/host-linux-x64/reports/cuda_gpu_mem_size_sum.py]...

 ** CUDA GPU MemOps Summary (by Size) (cuda_gpu_mem_size_sum):

 Total (MB)  Count  Avg (MB)  Med (MB)  Min (MB)  Max (MB)  StdDev (MB)      Operation
 ----------  -----  --------  --------  --------  --------  -----------  ------------------
      0.008      2     0.004     0.004     0.004     0.004        0.000  [CUDA memcpy HtoD]
      0.004      1     0.004     0.004     0.004     0.004        0.000  [CUDA memcpy DtoH]
 
```
::::

> ### Interpreting Key Outputs

Reviewer: ```suggestion
The provided nsys report contains key output that can help you optimize your code. For more information on these key output sections and interpreting the output, please review the tabs below.
```
For the other tools, we do not create a subsection for interpreting the output. So, I think we should remove it and replace it with my suggestion.
